2단계 - 문자열 덧셈 계산기

bestsilver opened the pull request:
안녕하세요!
2단계 문자열 덧셈 계산기 리뷰 요청 드립니다!

감사합니다 : )

bestsilver committed:
docs: 기능 요구사항 및 구현 기능 목록 작성

bestsilver committed:
chore: empty 파일들 삭제

bestsilver committed:
feat: 음수이거나 숫자 이외의 값이 넘어왔을 때 exception 구현

bestsilver committed:
feat: StringPattern (text의 종류 정의)

bestsilver committed:
feat: StringPatternRegex (text의 종류를 구분하기 위한 정규표현식들 정의)

bestsilver committed:
feat: Separator (text의 종류에 따라 separate할 수 있는 기능 구현)

bestsilver committed:
feat: Separator

- text의 종류 판별하는 기능
- text가 empty인지 판별하는 기능
- text의 종류에 따라 separate 후 요소를 sum하는 기능

bestsilver committed:
test: StringAddCalculatorTest

- empty 체크
- 숫자 하나 체크
- 쉼표 구분자
- 쉼표 또는 콜론 구분자
- custom 구분자
- 음수 체크

enemfk777 commented on src/main/java/calculator/model/StringAddCalculator.java:
```diff
@@ -0,0 +1,45 @@
+
+    public static int sumNumberList(String[] list) throws RuntimeException {
+        try{
+            return Arrays.stream(list).mapToInt(Integer::parseInt).sum();
```
"음수가 아닌" "정수" 라는 비즈니스 요구사항을 나타내는 책임을 갖는 클래스가 분리된다면 어떨까요??

bestsilver replied:
Numbers라는 클래스를 만들어 따로 분리해보았습니다!

enemfk777 commented on src/main/java/calculator/model/StringPatternRegex.java:
```diff
@@ -0,0 +1,6 @@
+package calculator.model;
+
+public class StringPatternRegex {
+    public static final String FIXEDPATTERN = "(\\d+(,|:)?)+";
+    public static final String CUSTOMPATTERN = "//(.)\n(.*)";
```
enum을 활용해 보면 어떨까요??

bestsilver replied:
StringPatternRegex 클래스를 삭제 후 StringPattern으로 통합했습니다!

enemfk777 commented on src/main/java/calculator/model/StringAddCalculator.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public static StringPattern determineStringPattern(String text) {
+        if (text.matches(StringPatternRegex.FIXEDPATTERN)) {
+            return StringPattern.FIXED;
+        }
+
+        if (text.matches(StringPatternRegex.CUSTOMPATTERN)) {
+            return StringPattern.CUSTOM;
+        }
```
StringPatternRegex가 enum이 된다면, 이 if 문도 사라질 수 있을것 같아요 :)

enemfk777 commented on src/main/java/calculator/model/Separator.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    private void initFixedPattern(String text) {
+        this.delimiter = ",|:";
```
",|:" 를 의미를 갖는 상수로 명명해보시는건 어떨까요? :)

bestsilver replied:
매직스트링이 있었군요 ㅠㅠ 상수로 선언해보았습니다!

enemfk777 commented on src/main/java/calculator/model/StringAddCalculator.java:
```diff
@@ -0,0 +1,45 @@
+public class StringAddCalculator {
+
+    public static int splitAndSum(String text) throws RuntimeException{
+        if (isEmpty(text)) {
```
판별 조건을 의미를 잘 나타낼 수 있는 메서드로 분리해주셔서 가독성이 좋은 코드가 되었네요 👍

enemfk777 commented on src/test/java/calculator/model/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertThatThrownBy(() -> StringAddCalculator.splitAndSum("-1,2,3"))
+                .isInstanceOf(RuntimeException.class);
+    }
+}
```
파일 끝은 개행으로 끝나는 것이 POSIX 표준이어서 깃허브에서 경고해주고 있네요 :)
아래는 관련 내용이 있는 글입니다..
https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/

아래는 인텔리제이에서 자동으로 파일 끝에 개행을 추가하는 옵션에 관한 글입니다
https://velog.io/@d-h-k/intellij-%ED%8C%8C%EC%9D%BC%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EC%B6%94%EA%B0%80%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95

bestsilver replied:
허..그렇군요 ㅠㅠ 적용했습니다!!
오..친절하게 해결 링크까지~!감사합니다!! 👍 👍

enemfk777 reviewed (changes requested):
안녕하세요 고은님 🙇‍♂️
문자열 덧셈 계산기를 잘 구현해주셨네요..!!
몇가지 코멘트 남겼으니 확인 후 다시 한번 PR 부탁드립니다
!!!
화이팅입니다 👍

bestsilver committed:
docs: refactor 후 기능 변경 사항

- Numbers 객체 추가
- StringPatternRegex 삭제, StringPattern으로 합치기

bestsilver committed:
refactor: StringPatternRegex 삭제, StringPattern으로 합치기

bestsilver committed:
feat: 덧셈을 위한 숫자 리스트를 가지는 클래스 구현
- sum 기능 구현
- 음수 포함하는지 체크하는 기능 구현

bestsilver committed:
refactor:

- 매직스트링 제거
- StringPatternRegex 삭제 후 StringPattern으로 통합 적용

bestsilver committed:
refactor:

- text의 매칭값에 따라 StringPattern을 리턴해주는 기능을 StringPattern 클래스 내부로 이동
- 정수 책임을 가지는 클래스로 분리하기 위해 Numbers 구현 후 분리

bestsilver committed:
refactor: 매직스트링 상수로 선언

enemfk777 commented on src/main/java/calculator/model/Separator.java:
```diff
@@ -0,0 +1,42 @@
+    private String text;
+    private static final String commaOrColonDelimiter = "[,:]";
+
+    Separator(StringPattern pattern, String text) throws RuntimeException {
+        if (pattern == StringPattern.FIXED) {
+            initFixedPattern(text);
+        }
+
+        if (pattern == StringPattern.CUSTOM) {
+            initCustomPattern(text);
+        }
+
+        if (pattern == StringPattern.NOTHING) {
+            throw IllegalArgument.NEGATIVE_OR_NOT_NUMBER;
+        }
+    }
```
FunctionalInterface를 활용한다면 이부분도 StringPattern enum 안에서 수행 될 수 있을 것 같아요 :)

enemfk777 commented on src/main/java/calculator/model/Numbers.java:
```diff
@@ -0,0 +1,28 @@
+import exception.IllegalArgument;
+import java.util.Arrays;
+
+public class Numbers {
```
일급 컬렉션을 잘 만들어 주셨네요! 💯 
Numbers 라는 일급 컬렉션이 있다면 Number라는 클래스도 존재 할 수 있을 것 같은데요
그렇다면 음수가 아닌지 확인하는 책임은 어디에 위치하는게 자연스러울까요? :)

enemfk777 commented on src/main/java/calculator/model/Numbers.java:
```diff
@@ -0,0 +1,28 @@
+        return Arrays.stream(numbers).anyMatch(n -> n < 0);
+    }
+
+    public int sumNumbers() {
```
새로 도출된 클래스들의 테스트코드도 작성된다면 더 단단한 코드가 될 것 같아요 :)

enemfk777 reviewed (approved):
코멘트 드린 부분 잘 반영해주셨네요 💯 

몇가지 코멘트 드린 부분은 슥 보시고 다음 스텝에서 반영해주시지 않아도 됩니다 :)
아마 오늘 강의에서 언급될 수 있는 내용일 거예요 :)

이번 단계는 이만 머지 하겠습니다..!
그럼 남은 단계도 화이팅입니다 🙏
